File: tutorial/java/Client/src/main/java/systems/zlink/tutorial/client/HttpOperationsController.java
package systems.zlink.tutorial.client;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;

import jakarta.servlet.http.HttpServletRequest;

import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.servlet.mvc.method.annotation.StreamingResponseBody;

import systems.zlink.framework.channels.ZLinkRouteClient;
import systems.zlink.tutorial.shared.Contracts;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.nio.charset.StandardCharsets;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;
import java.util.concurrent.CompletionStage;

@RestController
final class HttpOperationsController {

    private static final MediaType NDJSON = MediaType.valueOf("application/x-ndjson");

    private final ZLinkRouteClient route;
    private final ObjectMapper mapper;

    HttpOperationsController(ZLinkRouteClient route, ObjectMapper mapper) {
        this.route = route;
        this.mapper = mapper;
    }

    @GetMapping("/rooms/{roomId}/export")
    CompletionStage<ResponseEntity<StreamingResponseBody>> exportRoom(@PathVariable String roomId) {
        return route.requestToSpot(roomId, new Contracts.GetRoomState())
                .submit(Contracts.RoomState.class)
                .thenApply(state -> roomExport(roomId, state));
    }

    private ResponseEntity<StreamingResponseBody> roomExport(
            String roomId, Contracts.RoomState state) {
        return ResponseEntity.ok()
                .contentType(NDJSON)
                .body(output -> writeRoom(output, roomId, state));
    }

    private void writeRoom(OutputStream output, String roomId, Contracts.RoomState state)
            throws IOException {
        writeLine(output, new RoomLine(roomId, state.title(), null));
        for (String message : state.chat()) {
            writeLine(output, new RoomLine(null, null, message));
        }
    }

    @PostMapping("/rooms/{roomId}/import")
    CompletionStage<Imported> importRoom(@PathVariable String roomId, HttpServletRequest request)
            throws IOException {
        return readNext(
                new BufferedReader(
                        new InputStreamReader(request.getInputStream(), StandardCharsets.UTF_8)),
                roomId,
                0);
    }

    private CompletionStage<Imported> readNext(BufferedReader reader, String roomId, int count) {
        return CompletableFuture.supplyAsync(() -> readLine(reader))
                .thenCompose(
                        line -> {
                            if (line == null) {
                                close(reader);
                                return CompletableFuture.completedFuture(new Imported(count));
                            }
                            Contracts.PostChat message = parseChat(line);
                            return route.sendToSpot(roomId, message)
                                    .submit()
                                    .thenCompose(ignored -> readNext(reader, roomId, count + 1));
                        });
    }

    private static String readLine(BufferedReader reader) {
        try {
            return reader.readLine();
        } catch (IOException error) {
            throw new CompletionException(error);
        }
    }

    private Contracts.PostChat parseChat(String line) {
        try {
            return mapper.readValue(line, Contracts.PostChat.class);
        } catch (JsonProcessingException error) {
            throw new CompletionException(error);
        }
    }

    private void writeLine(OutputStream output, RoomLine value) throws IOException {
        output.write(mapper.writeValueAsBytes(value));
        output.write('\n');
        output.flush();
    }

    private static void close(BufferedReader reader) {
        try {
            reader.close();
        } catch (IOException error) {
            throw new CompletionException(error);
        }
    }

    private record RoomLine(String roomId, String title, String message) {}

    private record Imported(int imported) {}
}

File: tutorial/kotlin/Client/src/main/kotlin/systems/zlink/tutorial/client/HttpOperationsController.kt
package systems.zlink.tutorial.client

import com.fasterxml.jackson.databind.ObjectMapper
import jakarta.servlet.http.HttpServletRequest
import java.io.OutputStream
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.future.await
import kotlinx.coroutines.withContext
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.servlet.mvc.method.annotation.StreamingResponseBody
import systems.zlink.framework.channels.ZLinkRouteClient
import systems.zlink.tutorial.shared.GetRoomState
import systems.zlink.tutorial.shared.PostChat
import systems.zlink.tutorial.shared.RoomState

@RestController
class HttpOperationsController(route: ZLinkRouteClient, private val mapper: ObjectMapper) {

    private val route = route

    @GetMapping("/rooms/{roomId}/export")
    suspend fun exportRoom(@PathVariable roomId: String): ResponseEntity<StreamingResponseBody> {
        val state =
            route.requestToSpot(roomId, GetRoomState()).submit(RoomState::class.java).await()
        val body = StreamingResponseBody { output ->
            writeLine(output, RoomLine(roomId, state.title, null))
            state.chat.forEach { message -> writeLine(output, RoomLine(null, null, message)) }
        }
        return ResponseEntity.ok().contentType(MediaType.valueOf("application/x-ndjson")).body(body)
    }

    @PostMapping("/rooms/{roomId}/import")
    suspend fun importRoom(@PathVariable roomId: String, request: HttpServletRequest): Imported {
        val reader = request.inputStream.bufferedReader()
        var count = 0
        try {
            while (true) {
                val line = withContext(Dispatchers.IO) { reader.readLine() } ?: break
                val message = mapper.readValue(line, PostChat::class.java)
                route.sendToSpot(roomId, message).submit().await()
                count += 1
            }
        } finally {
            withContext(Dispatchers.IO) { reader.close() }
        }
        return Imported(count)
    }

    private fun writeLine(output: OutputStream, value: RoomLine) {
        output.write(mapper.writeValueAsBytes(value))
        output.write('\n'.code)
        output.flush()
    }

    private data class RoomLine(val roomId: String?, val title: String?, val message: String?)

    data class Imported(val imported: Int)
}
